Sidebar unification

GuillaumeGomez opened the pull request:
This PR does a few things:
 * Put crates list at all levels (before, it was only on the "top" items)
 * Fix bug in module sidebar: the list of items was from the parent module.

The other changes (on bootstrap mostly) were to allow to generate multiple crates in a same folder so that we can ensure that clicking on the crates in the sidebar works as expected.

I added a rustdoc-gui test to ensure everything is where it should be.

r? @jyn514

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2487765294) [(plain)](https://github.com/rust-lang/rust/commit/da3097bdc07547cf5111158e586e40eebe0bc075/checks/2487765294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
tidy: Skipping binary file check, read-only filesystem
* 623 error codes
* highest error code: E0781
tidy error: /checkout/src/tools/rustdoc-gui/tester.js:76: line longer than 100 chars
Found 0 error codes with no tests
Done!
* 330 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:12
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2487935285) [(plain)](https://github.com/rust-lang/rust/commit/f809c87e7b1b40204679bb943403a0b1095f9478/checks/2487935285/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11818
.................................................................................................... 9500/11818
.................................................................................................... 9600/11818
........................................i......i.................................................... 9700/11818
......................................................................................iiiiiii..iiiii 9800/11818
.................................................................................................... 10000/11818
.................................................................................................... 10100/11818
.................................................................................................... 10200/11818
.................................................................................................... 10300/11818
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 33 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii....
test result: ok. 4 passed; 0 failed; 29 ignored; 0 measured; 0 filtered out; finished in 0.11s

 finished in 0.179 seconds
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i...i.ii....i.i....ii............iiii.......i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.38s

 finished in 2.451 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
failures:

---- [rustdoc] rustdoc/keyword.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/keyword" "/checkout/src/test/rustdoc/keyword.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
7: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/index.html '//div[@class="block items"]//a/@href' '#keywords'
Encountered 1 errors

------------------------------------------

---
test result: FAILED. 433 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 31.15s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:15:10
```

</details>

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -725,40 +725,36 @@ function hideThemeButtonState() {
         var current = window.sidebarCurrent;
+        var isModule = getNakedUrl().endsWith("/index.html");
+
+        function addSidebarCrates(crates) {
```
I moved this function here so it has access to the `sidebar` variable from the scope instead of querying it like before.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -798,28 +794,31 @@ function hideThemeButtonState() {
-        block("keyword", "Keywords");
-        block("traitalias", "Trait Aliases");
+        if (sidebar) {
+            if (!isModule) {
```
We don't want to render these info on a module (they are already on listed on its page).

GuillaumeGomez commented:
@jyn514 This is now ready for review. I added a few comments on the changes to help you understand why I did the changes.

jyn514 commented:
I don't have time to review this.

r? @jsha

GuillaumeGomez commented:
It's actually not much a change in the front rather than in the back so maybe @camelid might be a better choice here (but any review is very welcome of course!).

r? @camelid

GuillaumeGomez commented:
And rebased too. :)

GuillaumeGomez commented:
And of course I pushed the wrong branch once again...

camelid commented:
Could you include a screenshot of before/after so I can get an idea of what has been changed or fixed?

camelid commented:
> fixed a bug where a module was listing items from the parent (funny that no one paid attention to it).

I don't think this is a bug; I think it is intended behavior. For example, say you have a crate like this:

```rust
pub mod sub1 {
    pub mod sub1sub1 {
        pub struct Sub1Sub1Struct;
    }

    pub mod sub1sub2 {
        pub struct Sub1Sub2Struct;
    }

    pub struct Sub1Struct;
}

pub mod sub2 {
    pub mod sub2sub1 {
        pub struct Sub2Sub1Struct;
    }

    pub struct Sub2Struct;
}

pub struct RootStruct;
```

Then, if I go to the `foo::sub1::sub1sub2` page, the sidebar looks like this:

![image](https://user-images.githubusercontent.com/37223377/117859556-023cee80-b244-11eb-9e6e-98e60572ed67.png)

The idea is that then I can click on `sub1sub2` or `Sub1Struct` and navigate to a sibling item. Notice that the header above 'Modules' and 'Structs' shows the path `foo::sub1`, which is the parent.

GuillaumeGomez commented:
This is very disturbing because then it's not coherent with when you're on the top module or any other pages. I really think this is much better with this PR.

GuillaumeGomez commented:
cc @rust-lang/rustdoc

Nemo157 commented:
The fact that the "Crates" section also appears on non-module items in the root seems like a bug to me.

GuillaumeGomez commented:
@Nemo157 It's part of the unification: why would only modules list other crates? I often found myself to look for it in item before going back to top to be able to switch.

GuillaumeGomez commented:
Arf, seems like we both have very different opinions on this haha.

bors commented:
:umbrella: The latest upstream changes (presumably #85199) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Stupremee commented on src/librustdoc/html/static/main.js:
```diff
@@ -564,40 +564,36 @@ function hideThemeButtonState() {
     window.initSidebarItems = function(items) {
         var sidebar = document.getElementsByClassName("sidebar-elems")[0];
         var current = window.sidebarCurrent;
+        var isModule = getNakedUrl().endsWith("/index.html");
```
This check might fail at the root module if the webserver redirects `/` to `/index.html`. In fact, this happens on doc.rust-lang.org causing https://doc.rust-lang.org/std/ to be the same as https://doc.rust-lang.org/std/index.html. 

This would cause `isModule` to be false, and lead to showing all items at the root level. You can reproduce it locally by running `python -m http.server`, which does this silent redirection.

GuillaumeGomez replied:
This is a good point, thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #85304) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

GuillaumeGomez commented:
Any opinion on this @rust-lang/rustdoc ?

camelid commented:
Has the content of this PR changed or is it still what https://github.com/rust-lang/rust/pull/84834#issue-628668245 says?

jsha commented:
I don't have a strong feeling for when the crates list in the sidebar is used. On doc.rust-lang.org, there are very few crates, and two of them are subsets of `std`. It seems like almost all your browsing gets done inside `std`. On docs.rs, each crate is a world unto itself, so there is no need for crates in the sidebar. So that leaves situations where people generate and host docs for a group of crates themselves. Is crate navigation in the sidebar used a lot for those?

If you're on a struct page with a typical number of methods an implementations, it seems likely that a list of crates would be pushed off the bottom of the sidebar anyhow. Is it more efficient to scroll to the bottom (potentially many pages worth) to get a crate list than it is to click the logo in the upper left?

For sibling modules I don't have a strong opinion but I see what @Nemo157 is getting at about siblings (or neighbors). One way to think about this: If you're looking at the list of methods in the sidebar, you can click one method, and then click another method without losing your place. In other words, selecting one item doesn't mean you lose the opportunity to select its neighbors. Similarly, it may be nice in some cases to select a module and not lose the chance to select its neighbors.

For instance, if you are looking at [std::collections::linked_list](https://doc.rust-lang.org/std/collections/linked_list/index.html), you can also quickly click on its neighbor module vec_deque.

<img width=200 src=https://user-images.githubusercontent.com/220205/118600781-4bba9b80-b766-11eb-9f38-1102943a6628.png>

But in general, I find that anything below the fold within the sidebar is not very useful for navigation. I think most users don't know what will be found below all the methods and implementations, and don't really think to go looking.

jsha commented:
One practical thing I like about @GuillaumeGomez's proposal: If you're looking at some second-level module in std, like [`std::env`](https://doc.rust-lang.org/std/env/index.html), the sidebar is full of a _lot_ of stuff that seems really irrelevant. There are 24 primitive types, 49 macros, and 39 keywords! It's definitely not clear why all those things are listed in the sidebar of all the second-level modules but not in the root or anywhere else. I think that's worth fixing like so:

Primitive types should only be listed in the sidebar of other primitive types
Keywords should only be listed in the sidebar of other keywords
Macros should only be listed in the sidebar of other macros

In general they should all be primarily listed as subheadings at https://doc.rust-lang.org/std/, as they currently are.

I think my general rule here is: we should really strive to make the sidebar not very long in most cases, and include in it only the most targeted and relevant stuff.

Nemo157 commented:
> On docs.rs, each crate is a world unto itself, so there is no need for crates in the sidebar.

This is something I would like to see change, we should document all crates in the package together https://github.com/rust-lang/docs.rs/issues/238.

GuillaumeGomez commented:
> Has the content of this PR changed or is it still what [#84834 (comment)](https://github.com/rust-lang/rust/pull/84834#issue-628668245) says?

I updated my first comment when I made the changes, so the first comment is up to date. For now I'm waiting for us to reach an agreement. But it seems like @jsha is agreeing with me for the moment, so already a nice start. :)

Nemo157 commented:
> It's definitely not clear why all those things are listed in the sidebar of all the second-level modules but not in the root or anywhere else.

`std` is special in this case that it doesn't have other "normal" items at the second-level. Normally the sidebar of any second-level item will show all other second-level items, e.g. <https://docs.rs/rayon/1.5.0/rayon/struct.FnContext.html> below the trait lists.

As to why they're not in the root, :shrug: for some reason modules only show links to their headings, not all their items. Even in cases like <https://doc.rust-lang.org/stable/std/boxed/index.html> where there is only a single item that would easily fit in the sidebar.

> Primitive types should only be listed in the sidebar of other primitive types
> Keywords should only be listed in the sidebar of other keywords
> Macros should only be listed in the sidebar of other macros

Would you apply this to other crates too? For example should <https://docs.rs/futures/0.3.15/futures/macro.pending.html> not show the "Modules" section?

Manishearth commented:
> I don't have a strong feeling for when the crates list in the sidebar is used. On doc.rust-lang.org, there are very few crates, and two of them are subsets of `std`. It seems like almost all your browsing gets done inside `std`.

I've definitely gotten feedback that `proc_macro` and some other libraries are not discoverable from std so we probably should have them in the sidebar and probably also link to them in other ways. But perhaps the solution here is just to link to them prominently elsewhere in the module docs.

GuillaumeGomez commented:
With this PR, the number of items in the modules' sidebar is strongly reduced, making it much easier to see the crate list (no need to scroll in my resolution for example).

GuillaumeGomez commented:
I fixed @Stupremee comment: now I simply check if the body has the `mod` class to determine if it's a module (crate module also has this class).

So with this, I think it's good?

r? @jsha

bors commented:
:umbrella: The latest upstream changes (presumably #85838) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

jsha commented:
Linking in some related issues: #85759, #16328, #81031. The theme of these is that some people think the list of crates in the sidebar is much too long. For instance, I just did a quick check on `reqwest`: it has 71 crates in its sidebar. One of the linked issues mentions servo had 326 in 2018. Adding the crates list to every page would make that problem worse (and would increase page sizes somewhat, for a very rarely used navigational element).

The more I think about it the more I agree with what @Nemo157 said:

> the behavior of the "Crates" section appearing anywhere other than at the root of the crate seems like a bug to me, and the fix would be to reduce the places it appears, not adding more.

(BTW sorry if I gave the impression earlier that I fully support the approach here; I'm still thinking aloud about how I think this _should_ work).

We have a hierarchical structure of pages:

```
crate1
crate2
  struct1
  func1
  moduleA
    struct2
    moduleB
      struct3
    moduleC
      struct4
      struct5
```

And the question is, at any given point in that structure, what navigation options should we offer in the sidebar? The minimalist view is "breadcrumb" style: from struct4, you have direct links to go to moduleC, moduleA, crate2, or the doc index page. If you want to get from struct4 to struct3, you need to go up to moduleA, then down to moduleB. We have breadcrumbs in place already at the top of the page: in `Struct foo::bar::Baz`, you can click on foo or bar.

A medium approach would be "siblings" style: from any given node in the hierarchy, you have links to go to all siblings. So struct4 has a link to struct5; moduleB has a link to moduleC; crate1 has a link to crate1.

A maximal approach would be "siblings of all parents:" from any given node, you have links to all siblings, and also the siblings of all your parent nodes.

Right now I think we have a bit of a mix. If you're at struct1, the sidebar will list func1 (a sibling) and moduleA (a sibling), but also crate1 (a parent's sibling).

After writing all that, here's what I think would be consistent:
 - Sidebar is for siblings; document body is for children. In particular module pages contain their sister modules in the sidebar, and their child modules and types in the document body.
 - Crates are siblings to each other, so each crate page includes all the crates in its sidebar. Non-crate pages never have crates in their sidebar.

I would also say we have a general problem that the logic of the sidebar is not discoverable. For instance, https://docs.rs/rustls/0.19.1/rustls/struct.DangerousClientConfig.html. The first part of the sidebar is "things that exist on this page." Then it says `rustls` in a white box and a bunch more items below. It's not very clear what that white box means. It's supposed to mean "we're done with items from this page; below are some items from elsewhere in this crate." It's also not clear how comprehensive that list is: Under "Structs" am I looking at all the structs in this docset? In this crate? (It's in this module, of course). Perhaps one simple fix would be for the white box to say "Other items in rustls."

jsha commented:
Here's a small prototype of what I think should happen for the crates listing. It uses a regex to determine "on a crate page;" a real version would do better. https://github.com/rust-lang/rust/compare/master...jsha:crates-have-neighbors?expand=1

Demo of generated docs for hyper:

https://hoffman-andrews.com/rust/crates-have-neighbors-hyper/hyper/index.html

Demo of generated docs for library/std:

https://hoffman-andrews.com/rust/crates-have-neighbors/std/

GuillaumeGomez commented:
I'm a bit sad about the crates being only listed at the top level but if I'm the only one finding it illogical and annoying to go to the crate level to have access to it, then so be it.

> Perhaps one simple fix would be for the white box to say "Other items in rustls."

This is a good idea!

> It uses a regex to determine "on a crate page;"

I'll instead add a new `crate` class on the `<body>` (so the body will have both `mod` and `crate` classes) to make things simpler.

GuillaumeGomez commented:
So I did the following:

 * crate list is now only listed on the crate page.
 * I added a new `crate` class on the `<body>` of the crate page to make it simpler to check if we are on a crate page.
 * I used @jsha's suggestion for improving the separation between the page's items and siblings (screenshot below)
 * I added underline when we hover the link in the separation too (to make it more coherent with what we have everywhere else).

![Screenshot from 2021-05-31 11-57-59](https://user-images.githubusercontent.com/3050060/120178669-08ead000-c20a-11eb-9fb8-97fdd0821960.png)

jyn514 commented on src/librustdoc/html/render/context.rs:
```diff
@@ -200,8 +200,15 @@ impl<'tcx> Context<'tcx> {
         let keywords = make_item_keywords(it);
+        let name;
+        let tyname_s = if it.is_crate() {
+            name = format!("{} crate", tyname.as_str());
```
```suggestion
            name = format!("{} crate", tyname);
```

GuillaumeGomez replied:
Good catch!

jsha commented on src/tools/rustdoc-gui/tester.js:
```diff
@@ -63,6 +63,13 @@ async function main(argv) {
         // This is more convenient that setting fields one by one.
         options.parseArguments([
             "--no-screenshot",
+            // This option shows what puppeteer "code" is run
+            // "--debug",
+            // This option disable the headless mode, allowing you to see what's going on.
+            // "--no-headless",
+            // The text isn't rendered by default because of a lot of small differences
+            // between hosts.
+            // "--show-text",
```
Is this intended to be here but commented out?

GuillaumeGomez replied:
Absolutely. Both options are very useful when adding/debugging a test. I have added and removed them a few times already.

jsha commented:
@bors r+

Looks good! 🎉

bors commented:
:pushpin: Commit 9b637fa72e4d6a3aebcbf92cc4e18b94a7bf31b7 has been approved by `jsha`

<!-- @bors r=jsha 9b637fa72e4d6a3aebcbf92cc4e18b94a7bf31b7 -->
<!-- homu: {"type":"Approved","sha":"9b637fa72e4d6a3aebcbf92cc4e18b94a7bf31b7","approver":"jsha"} -->

bors commented:
:hourglass: Testing commit 9b637fa72e4d6a3aebcbf92cc4e18b94a7bf31b7 with merge 19579c656436d7998289399ca373889b8e6019ce...
<!-- homu: {"type":"BuildStarted","head_sha":"9b637fa72e4d6a3aebcbf92cc4e18b94a7bf31b7","merge_sha":"19579c656436d7998289399ca373889b8e6019ce"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2734055301)
Approved by: jsha
Pushing 19579c656436d7998289399ca373889b8e6019ce to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jsha","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2734055301"},"merge_sha":"19579c656436d7998289399ca373889b8e6019ce"} -->
